### PR TITLE
vector: 0.10.0 -> 0.11.0

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

upgrade vector

###### Things done

This is more a RFC at this point. @thoughtpolice when you have a moment.

(I've run nixfmt on this, I hope it's not a problem. The only change is version and sha256 corresponding)

Trying to build this, you'll get the 
```
error: failed to sync

Caused by:
  found duplicate version of package `tracing-attributes v0.1.11` vendored from two sources:

        source 1: registry `https://github.com/rust-lang/crates.io-index`
        source 2: https://github.com/tokio-rs/tracing?rev=f470db1b0354b368f62f9ee4d763595d16373231#f470db1b
Traceback (most recent call last):
  File "/nix/store/jfq4wq19ms0m46nkwr9xf3kiv9gqs2jl-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 42, in <module>
    main()
  File "/nix/store/jfq4wq19ms0m46nkwr9xf3kiv9gqs2jl-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 17, in main
    assert list(data.keys()) == ["source"]
AssertionError
```

This is caused by the fact that vector has a dependency on two libraries named tracing with different sources and versions.
https://github.com/timberio/vector/blob/master/Cargo.lock#L6717
and
https://github.com/timberio/vector/blob/master/Cargo.lock#L6727

I'm not sure what's the best way to resolve this?

In this case, vector has all it's dependencies defined clearly in the cargo.lock, is there even a need to run the vendored script?

@symphorien @Mic92 perhaps you have a better idea of how to solve the problem of having two dependencies named exactly the same?

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/vector/default.nix [hunk 1]
```diff
@@ -1,42 +1,53 @@
...
+  "jemallocator"
+  "rdkafka"
+  "rdkafka/dynamic_linking"
+] else [
+  "leveldb"
+  "leveldb/leveldb-sys-2"
+  "jemallocator"
+  "rdkafka"
+  "rdkafka/dynamic_linking"
+]) ++ (lib.optional stdenv.targetPlatform.isUnix "unix")
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/vector/default.nix:14]:
```suggestion
, features ? ((if !stdenv.isAarch64 then [
  "leveldb"
  "leveldb/leveldb-sys-2"
]) ++ [
  "jemallocator"
  "rdkafka"
  "rdkafka/dynamic_linking"
] ++ (lib.optional stdenv.targetPlatform.isUnix "unix")
```

Did not test.

File: pkgs/tools/misc/vector/default.nix [hunk 2]
```diff
@@ -1,42 +1,53 @@
...
-    rev    = "v${version}";
-    sha256 = "0q6x3fvwwh18iyznqlr09n3zppzgw9jaz973s8haz54hnxj16wx0";
+    owner = "timberio";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "xyk/6WYCzD1IM/jgt+6VT2GTNClE8qdUSEXWQlpY/Bg=";
   };
 
-  cargoSha256 = "Y/vDYXWQ65zZ86vTwP4aCZYCMZuqbz6tpfv4uRkFAzc=";
+  cargoSha256 = "";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/vector/default.nix:28]: You need to set this.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/vector/default.nix (hunk 2 above)
Comment (happysalada) [reply pkgs/tools/misc/vector/default.nix:28]: Agreed. I wanted to make it clear that this needs to be updated. But you are perfectly right!
I'll leave this here until the build issue has been solved.
Thanks!

**Comment by kvtb** [comment]
`vector` since `0.11.0` has wrong `Cargo.lock`.
It is strange that nixpkgs' Rust pipeline rely on `Cargo.lock` instead on `Cargo.toml`

**Comment by kvtb** [comment]
also related https://github.com/rust-lang/cargo/issues/7285

**Comment by happysalada** [comment]
Thanks for the comment! I'm guessing this will not be possible to merge in the short term then.
Perhaps I should leave this open for now until something changes.

**Comment by symphorien** [comment]
Technically, if you can obtain a "correct" cargo.lock by pinning versions differently, you can use this cargo.lock (see section 15.23.1.5. Building a crate with an absent or out-of-date Cargo.lock file of the nixpkgs manual)

**Comment by happysalada** [comment]
@symphorien Thanks for the update
Here is a link to the [rust part of the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#rust)

I've just checked the section, and they suggest using cargo.lock patch. However in this case I'm not even sure what to patch, as the cargo.lock is potentially correct, just referring to two different crates with the same name.
If anybody has an idea of what to patch, I'm open to trying it.

**Comment by SuperSandro2000** [comment]
@happysalada please fix the eval error.

**Comment by happysalada** [comment]
@SuperSandro2000 thanks for checking up on this.
I don't know how to fix the eval error. I was hoping for some outside help.
The problem is that there are two packages with the same name and different versions. The cargo vendor step done in nixpkgs doesn't like that.
I don't mind just closing the PR if it's better to not leave it open.

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 thanks for checking up on this.
> I don't know how to fix the eval error. I was hoping for some outside help.
> The problem is that there are two packages with the same name and different versions. The cargo vendor step done in nixpkgs doesn't like that.
> I don't mind just closing the PR if it's better to not leave it open.

Oh, didn't read the rest. I would create an issue how upstream wants to support cargo vendor.

**Comment by happysalada** [comment]
the issue was opened upstream. So this is waiting for it. Let me know if anything else is needed.

**Comment by Mic92** [comment]
I have seen this issue with cargo-vendor before but also don't have a solution for it.

**Comment by kvtb** [comment]
> I have seen this issue with cargo-vendor before but also don't have a solution for it.

`cargo update --aggressive` before `cargo vendor` might help

from https://github.com/timberio/vector/issues/5738#issuecomment-766714994, which is one of the upstream issues triggered by our problem

**Comment by symphorien** [comment]
cargo update is not deterministic, as its result depends on network.

**Comment by kvtb** [comment]
> cargo update is not deterministic, as its result depends on network.

yes, it might result in introduction of another `cargoSha256`

**Comment by happysalada** [comment]
superseded

**Pull request closed without merging**